basic/airdrop-with-withdraw/src/lib.rs

```diff
@@ -0,0 +1,83 @@
+use scrypto::prelude::*;
+
+
+blueprint! {
+
+    struct AirdropWithWithdraw {
+        admin_badge: Address,
+        recipients: HashMap<Address, Decimal>,
+        tokens : Vault
+    }
+
+    impl AirdropWithWithdraw {
+        
+        pub fn new(token : Address) -> (Component, Bucket) {
+
+            let admin_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+                                .initial_supply_fungible(1);
+
+
+            
+            let component = Self {
+                admin_badge : admin_badge.resource_address(),
+                recipients :  HashMap::new(),
+                tokens : Vault::new(token),
+            }
+            .instantiate();
+
+            return (component, admin_badge)
+        }
+
+        fn get_required_tokens_internal(&self) -> Decimal {
+            let mut required_tokens  = Decimal::zero();
+            for val in self.recipients.values() {
+                required_tokens = required_tokens + *val ;  
+            }
+            return required_tokens;
+        }
+
+        #[auth(admin_badge)]
+        pub fn get_required_tokens(&self) -> Decimal {
+            return self.get_required_tokens_internal();
+        }
+
+        #[auth(admin_badge)]
+        pub fn add_recipient(&mut self, recipient: Address, tokens : Decimal)  {
```

Comment:
You could accept a `Bucket` here instead of Decimal. That way, you can update the `recipients` map and `tokens` vault in the same method